common/src/main/java/org/apache/nemo/common/DirectByteBufferOutputStream.java

```diff
@@ -0,0 +1,163 @@
+   */
+  @Override
+  public void write(final int b) {
+    ByteBuffer currentBuf = (dataList.isEmpty() ? null : dataList.getLast());
+    if (currentBuf == null || currentBuf.remaining() <= 0) {
+      newLastBuffer();
+      currentBuf = dataList.getLast();
+    }
+    currentBuf.put((byte) b);
+  }
+
+  /**
+   * Writes {@code b.length} bytes from the specified byte array
+   * to this output stream.
+   *
+   * @param b the byte to be written.
+   */
+  @Override
+  public void write(final byte[] b) {
+    write(b, 0, b.length);
+  }
+
+  /**
+   * Writes {@code len} bytes from the specified byte array
+   * starting at offset {@code off} to this output stream.
+   *
+   * @param   b     the data.
+   * @param   off   the start offset in the data.
+   * @param   len   the number of bytes to write.
+   */
+  @Override
+  public void write(final byte[] b, final int off, final int len) {
+    int byteToWrite = len;
+    int offset = off;
+    ByteBuffer currentBuf = (dataList.isEmpty() ? null : dataList.getLast());
+    while (byteToWrite > 0) {
+      if (currentBuf == null || currentBuf.remaining() <= 0) {
+        newLastBuffer();
+        currentBuf = dataList.getLast();
+      }
+      final int bufRemaining = currentBuf.remaining();
+      if (bufRemaining < byteToWrite) {
+        currentBuf.put(b, offset, bufRemaining);
+        offset += bufRemaining;
+        byteToWrite -= bufRemaining;
+      } else {
+        currentBuf.put(b, offset, byteToWrite);
+        offset += byteToWrite;
+        byteToWrite = 0;
+      }
+    }
+  }
+
+  /**
+   * Creates a byte array that contains the whole content
+   * currently written in this output stream.
+   *
+   * @return the current contents of this output stream, as byte array.
+   */
+  public byte[] toByteArray() {
+    if (dataList.isEmpty()) {
+      final byte[] byteArray = new byte[0];
+      return byteArray;
+    }
+
+    ByteBuffer lastBuf = dataList.getLast();
+    // pageSize equals the size of the data filled in the ByteBuffers
+    // except for the last ByteBuffer. The size of the data in the
+    // ByteBuffer can be obtained by calling ByteBuffer.position().
+    final int arraySize = pageSize * (dataList.size() - 1) + lastBuf.position();
+    final byte[] byteArray = new byte[arraySize];
+    int start = 0;
+    int byteToWrite;
+
+    for (ByteBuffer temp : dataList) {
+      // ByteBuffer has to be shifted to read mode by calling ByteBuffer.flip(),
+      // which sets limit to the current position and sets the position to 0.
+      // Note that capacity remains unchanged.
+      temp.flip();
+      byteToWrite = temp.remaining();
+      temp.get(byteArray, start, byteToWrite);
+      start += byteToWrite;
+    }
+    // The limit of the last buffer has to be set to the capacity for re-write.
+    lastBuf.limit(lastBuf.capacity());
+
+    return byteArray;
+  }
+
+  /**
+   * Closes this output stream and releases resources.
+   */
+  @Override
+  public void close() {
```

Comment:
The Outputstream API says that   
``` 
it closes the output stream. A closed stream cannot perform output operations and cannot be reopened.
```
but, just clearing the list doesn't guarantee this property

Reply:
I also wondered how I could actually 'close' this output stream, then I took a look at `ByteArrayOuputStream`'s `close()` in java.io:
>Closing a ByteArrayOutputStream has no effect. The methods in this class can be called after the stream has been closed without generating an IOException.

The purpose of this output stream is to substitute `DirectByteArrayOuputStream` that extends `ByteArrayOutputStream` so I guessed following this convention is okay. Am I on the right track?

Reply:
Then, could you please describe this in the comments? also if you follow this convention, clearing the list is necessary?

Reply:
I will add comments, thank you. The ByteBuffers in the list occupies the off-heap memory when allocated, and releases the memory when there is no reference to the ByteBuffer. This `close()` is intended to release the off-heap memory that the ByteBuffers occupy.

Reply:
I think it would be better to follow the convention of `OutputStream`, even though `ByteArrayOutputStream` does not do anything inside `close()`. One possible solution is maintaining a `isClosed` variable, and throws `IllegalStateException` on every read or write call if `isClosed` is `true`.

Reply:
I think that will be more straightforward. Thank you @taegeonum, @DifferentSC !

Reply:
@hy00nc Closing should not change the `toByteArray()` behavior, because we could want to retrieve `toByteArray()` after closing the stream. I think this is why `ByteArrayOutputStream` does nothing in `close()`

Reply:
@taegeonum It means that as long as this output stream implements `toByteArray()` it should also do nothing in `close()` including `clear()` for `LinkedList`, right?

Reply:
At a high level, i'm wondering why do we need `toByteArray` method? it returns on-heap byte array, but this class is for off-heap memory

Reply:
@taegeonum `toByteArray()` is still necessary because the data needs to be loaded into JVM heap from ByteBuffer using `get()` method before it is processed by JVM function. Storing array output stream outside JVM heap would still reduce the on-heap memory usage, because it will not be stored inside JVM memory until it's read.

BTW, we can implement some tricks to process off-heap data directly (e.g. off-heap hashing, zero-copy network transfer, ...), which removes the necessity for `toByteArray` as you say. This would be one of our future topics, and we still need `toByteArray` because those tricks can only be applied to some specific usages.

Reply:
@taegeonum According to the `java.io` documentation, `close()` should release any system resources associate with the `OutputStream` and after being `close()`, the `OutputStream` cannot perform any output operations. For this reason, I think we should remove the references for all the `ByteBuffer`, which makes it eligible for GC after being closed.

Reply:
@DifferentSC Yeah, I already pointed out the `close` API, but @hy00nc said that this class follows `DirectByteArrayOutputStream` semantic, which does nothing in `close`. So, my first question was that if you follow the `DirectByteArrayOutputStream`, then we should not do anything in `close`. 
After closing the stream, can't we retrieve the data? and is this the correct behavior? 

Also, if it serializes data but `toByteArray()` is called immediately, the performance will be degraded instead of using heap memory. So, this class assumes the specific pattern that reading the data will happen  after a long time. Then, I think the comments should include this assumption 

My suggestion: 
How about creating `getByteBuffer` method, which returns the current byte buffers, instead of `toByteArray`? 
then we can maybe implement `ByteBufferInputStream` that retreives bytes from the byte buffers.

Reply:
Avro also implemented `ByteBufferOutputStream`, which returns the list of `ByteBuffer`. 
https://avro.apache.org/docs/1.5.4/api/java/org/apache/avro/util/ByteBufferOutputStream.html